lib/puma/server.rb

```diff
@@ -344,12 +343,33 @@ def handle_check
       return false
     end
 
+    # Runs the provided block if the Reactor queue is enabled.
+    # Returns block return value if queue is enabled, or nil if queue is not enabled.
+    def if_queue_requests
+      @queue_mutex.synchronize do
+        if @queue_requests
+          yield
+        end
+      end
+    end
+    private :if_queue_requests
+
+    # Queues the provided client
+    def queue_client(client, timeout)
+      if_queue_requests do
+        client.set_timeout timeout
+        @reactor.add client
```

Comment:
Would it be possible to keep concurrency as a concern of the Reactor rather than the Server? It sort of feels like we could refactor this a little further to keep concurrency concerns out of this class.